src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -0,0 +1,318 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal partial class SymbolTreeInfo
+    {
+        private static SimplePool<MultiDictionary<string, MetadataDefinition>> s_definitionMapPool =
+            new SimplePool<MultiDictionary<string, MetadataDefinition>>(() => new MultiDictionary<string, MetadataDefinition>());
+
+        private static MultiDictionary<string, MetadataDefinition> AllocateDefinitionMap()
+        {
+            return s_definitionMapPool.Allocate();
+        }
+
+        private static void FreeDefinitionMap(MultiDictionary<string, MetadataDefinition> definitionMap)
+        {
+            definitionMap.Clear();
+            s_definitionMapPool.Free(definitionMap);
+        }
+
+        /// <summary>
+        /// this gives you SymbolTreeInfo for a metadata
+        /// </summary>
+        public static async Task<SymbolTreeInfo> GetInfoForMetadataReferenceAsync(
+            Solution solution,
+            PortableExecutableReference reference,
+            bool loadOnly,
+            CancellationToken cancellationToken)
+        {
+            var metadata = reference.GetMetadata();
+            if (metadata == null)
+            {
+                return null;
+            }
+
+            // Find the lock associated with this piece of metadata.  This way only one thread is
+            // computing a symbol tree info for a particular piece of metadata at a time.
+            var gate = s_metadataIdToGate.GetValue(metadata.Id, s_metadataIdToGateCallback);
+            using (await gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                SymbolTreeInfo info;
+                if (s_metadataIdToInfo.TryGetValue(metadata.Id, out info))
+                {
+                    return info;
+                }
+
+                // We don't include internals from metadata assemblies.  It's less likely that
```

Comment:
This comment seems misplaced now.  There is no mention of symbol accessibility here.